### PR TITLE
GemJoin7 adapter for USDT

### PR DESCRIPTION
**Pull request by wilbarnes**
amalgamation of GemJoin2/5 with implementation check

### CODE REVIEW

**Review by andy8052 — APPROVED** [review]
All tests passing for me. Looks good

File: src/join.sol [hunk 1]
```diff
@@ -427,3 +427,113 @@ contract AuthGemJoin is LibNote {
...
+
+    constructor(address vat_, bytes32 ilk_, address gem_) public {
+        gem = GemLike7(gem_);
+        dec = gem.decimals();
+        require(dec < 18, "GemJoin7/decimals-18-or-higher");
+        wards[msg.sender] = 1;
+        live = 1;
+        vat = VatLike(vat_);
+        ilk = ilk_;
+        setImplementation(address(gem), 1);
```
Comment (brianmcmichael) [inline src/join.sol]: This adds the `gem` address to the implementations array (currently `
0xdac17f958d2ee523a2206206994597c13d831ec7`), but it will never get checked.

You'd want to do `setImplementation(gem.upgradedAddress(), 1);` here. Even though it's currently `0x000...` we'd still want it to be approved.

File: src/join.sol [hunk 2]
```diff
@@ -427,3 +427,113 @@ contract AuthGemJoin is LibNote {
...
+        require(live == 1, "GemJoin7/not-live");
+        // mul does overflow check so require(wad < 2 ** 255) not needed
+        uint wad18 = mul(wad, 10 ** (18 - dec));
+        require(int(wad18) >= 0, "GemJoin7/overflow");
+
+        // implementation check
+        // tether uses a `deprecated` boolean; if deprecated is true, calls are forwarded to
+        // an `upgradedAddress` address. so check if deprecated, if so, require upgradedAddress
+        // to be approved by governance
+        if (gem.deprecated()) {
```
Comment (brianmcmichael) [inline src/join.sol]: I'm not sure this is necessary. As long as we set `implementations[0x000...] == 1` we'd only need to do one external load. If they ever do upgrade then this would perform 2 external loads on every call.

File: src/join.sol [hunk 3]
```diff
@@ -427,3 +427,113 @@ contract AuthGemJoin is LibNote {
...
+    function exit(address guy, uint wad) public note {
+        // mul does overflow check so require(wad < 2 ** 255) not needed
+        uint wad18 = mul(wad, 10 ** (18 - dec));
+        require(int(wad18) >= 0, "GemJoin5/overflow");
+
+        // implementation check
+        // tether uses a `deprecated` boolean; if deprecated is true, calls are forwarded to
+        // an `upgradedAddress` address. so check if deprecated, if so, require upgradedAddress
+        // to be approved by governance
+        if (gem.deprecated()) {
```
Comment (brianmcmichael) [inline src/join.sol]: Same as above.

File: src/join.sol [hunk 4]
```diff
@@ -427,3 +427,113 @@ contract AuthGemJoin is LibNote {
         require(gem.transfer(usr, wad), "AuthGemJoin/failed-transfer");
     }
 }
+
+interface GemLike7 {
+    // matches ERC20 spec
+    function decimals() external view returns (uint);
+    // matches ERC20 spec
```
Comment (brianmcmichael) [inline src/join.sol]: This is misleading, it's not a match because it doesn't return `bool`

File: src/join.sol [hunk 5]
```diff
@@ -427,3 +427,113 @@ contract AuthGemJoin is LibNote {
...
+    function transfer(address,uint) external;
+    // matches ERC20 spec
+    function transferFrom(address,address,uint) external;
```
Comment (brianmcmichael) [inline src/join.sol:439]: Also nonstandard

File: src/join.sol [hunk 6]
```diff
@@ -427,3 +427,113 @@ contract AuthGemJoin is LibNote {
...
+        );
+        require(ok, "GemJoin7/failed-transfer");
+
+        require(prevBalance - wad == gem.balanceOf(msg.sender), "GemJoin7/failed-transfer");
+    }
+
+    function exit(address guy, uint wad) public note {
+        // mul does overflow check so require(wad < 2 ** 255) not needed
+        uint wad18 = mul(wad, 10 ** (18 - dec));
+        require(int(wad18) >= 0, "GemJoin5/overflow");
```
Comment (brianmcmichael) [inline src/join.sol]: Update response to `GemJoin7/overflow`

File: src/DssDeploy.t.sol [hunk 7]
```diff
@@ -904,6 +924,54 @@ contract DssDeployTest is DssDeployTestBase {
...
+    }
+
+    function testFailGemJoin7Exit() public {
+        DSValue pip = new DSValue();
+        USDT usdt = new USDT(100 ether);
+        GemJoin7 usdtJoin = new GemJoin7(address(vat), "USDT", address(usdt));
+        dssDeploy.deployCollateral("USDT", address(usdtJoin), address(pip));
+        usdt.approve(address(usdtJoin), uint(-1));
+        usdtJoin.join(address(this), 10 * 10 ** 6);
+        usdt.deprecate(address(0));
```
Comment (brianmcmichael) [inline src/DssDeploy.t.sol]: I would choose a different address. The `upgradedAddress` is `0x000...` by default, so setting it to zero address here doesn't change it.

**Review by brianmcmichael — CHANGES_REQUESTED** [review]
See notes.

File: src/DssDeploy.t.sol [hunk 8]
```diff
@@ -904,6 +924,54 @@ contract DssDeployTest is DssDeployTestBase {
...
+    function testFailGemJoin7Join() public {
+        DSValue pip = new DSValue();
+        USDT usdt = new USDT(100 * 10 ** 6);
+        GemJoin7 usdtJoin = new GemJoin7(address(vat), "USDT", address(usdt));
+        dssDeploy.deployCollateral("USDT", address(usdtJoin), address(pip));
+        usdt.approve(address(usdtJoin), uint(-1));
+        assertEq(usdt.balanceOf(address(this)), 100 * 10 ** 6);
+        assertEq(usdt.balanceOf(address(usdtJoin)), 0);
+        assertEq(vat.gem("USDT", address(this)), 0);
+        usdt.deprecate(address(0));
```
Comment (brianmcmichael) [inline src/DssDeploy.t.sol]: Same note as on `testFailGemJoin7Exit`

File: src/join.sol (hunk 1 above)
Comment (wilbarnes) [reply src/join.sol]: resolved https://github.com/makerdao/dss-deploy/pull/53/commits/53c06b3097d6f8513b6dd7447df44b029fbb3105

File: src/join.sol (hunk 3 above)
Comment (wilbarnes) [reply src/join.sol]: resolved https://github.com/makerdao/dss-deploy/pull/53/commits/53c06b3097d6f8513b6dd7447df44b029fbb3105

removed if block

File: src/join.sol (hunk 4 above)
Comment (wilbarnes) [reply src/join.sol]: removed the comments

File: src/join.sol (hunk 5 above)
Comment (wilbarnes) [reply src/join.sol:439]: removed

File: src/join.sol (hunk 6 above)
Comment (wilbarnes) [reply src/join.sol]: done

File: src/DssDeploy.t.sol (hunk 7 above)
Comment (wilbarnes) [reply src/DssDeploy.t.sol]: resolved https://github.com/makerdao/dss-deploy/pull/53/commits/53c06b3097d6f8513b6dd7447df44b029fbb3105 

address(1)

File: src/DssDeploy.t.sol (hunk 8 above)
Comment (wilbarnes) [reply src/DssDeploy.t.sol]: resolved https://github.com/makerdao/dss-deploy/pull/53/commits/53c06b3097d6f8513b6dd7447df44b029fbb3105

address(1)

Comment (wilbarnes) [reply src/DssDeploy.t.sol]: resolved https://github.com/makerdao/dss-deploy/pull/53/commits/53c06b3097d6f8513b6dd7447df44b029fbb3105

address(1)

File: src/join.sol (hunk 2 above)
Comment (wilbarnes) [reply src/join.sol]: removed the if block

File: src/join.sol [hunk 9]
```diff
@@ -427,3 +427,93 @@ contract AuthGemJoin is LibNote {
...
+        implementations[implementation] = permitted; // 1 live, 0 distable
+    }
+
+    function mul(uint x, uint y) internal pure returns (uint z) {
+        require(y == 0 || (z = x * y) / y == x, "GemJoin7/overflow");
+    }
+
+    function join(address urn, uint wad) public note {
+        require(live == 1, "GemJoin7/not-live");
+        // mul does overflow check so require(wad < 2 ** 255) not needed
```
Comment (gbalabasquer) [inline src/join.sol]: Hmm not sure exactly what this comment means. The `< 2 ** 255` is for checking if a `uint` can be cast to `int`, which happens with the `int(wad18) >= 0`.
The `mul` itself checks a `uint` overflow but not an `int` one.

Comment (wilbarnes) [reply src/join.sol]: the `require(int(wad18) >= 0, GemJoin7/overflow")` checks the int overflow, yes? and the `mul()` checks the `uint` overflow, so there are sufficient checks in place?

i suppose my comment was more me thinking thru the logic

**Comment by gbalabasquer** [comment]
I think I'd prefer to simplify this adapter.
Just doing token.transfer() and transferFrom without being wrapped in a `call` and checking the boolean. It is actually pointless doing so, we didn't think it enough when we made `GemJoin2`.

On the other hand, checking the balances is kind of the same thing, it is not bad to add it but then we should add it for all the tokens (regardless the return value).

So in other words I think we could make it like `GemJoin6` but without the require in the transfer functions. And honestly `GemJoin6` should be just that, the same as it is but without the requires. The only case the require make sense is for tokens that have a return false instead of a revert in their implementation. Haven't seen one of them so far.

**Comment by gbalabasquer** [comment]
Are we actually sure USDT is not returning anything? as the return are in the implementation but not in the signature. We might even use the same GemJoin6.

**Comment by gbalabasquer** [comment]
Just pushed a few changes. The important ones are adding safe math for the calculation of `wad` in `GemJoin8` and moving the `int` casting check after the new value is calculated.

**Comment by gbalabasquer** [comment]
Btw I think `GemJoin7` should also implement the balance check as USDT can have fees.

**Comment by wilbarnes** [comment]
i shouldn't have added GemJoin8 to this repo, i was researching how to handle fees (and GemJoin8 is for PAXG which isn't until next month)

i will move the balance check to GemJoin7

**Comment by gbalabasquer** [comment]
Ahh ok, sounds good.

**Comment by wilbarnes** [comment]
two of the GemJoin6 tests, `testFailGemJoin6Join()` & `testFailGemJoin6Exit()` were missing `deployKeepAuth()` and were not testing what they were intended to, so i added that function call

**Comment by wilbarnes** [comment]
please review gemjoin7 adapter again and the `testGemJoin7FeeChange()` test, i added a `changeFees(,)` function in the tether mock token to adjust the fees

File: src/DssDeploy.t.sol [hunk 10]
```diff
@@ -904,6 +926,81 @@ contract DssDeployTest is DssDeployTestBase {
...
+        DSValue pip = new DSValue();
+        USDT usdt = new USDT(100 * 10 ** 6);
+        GemJoin7 usdtJoin = new GemJoin7(address(vat), "USDT", address(usdt));
+        dssDeploy.deployCollateral("USDT", address(usdtJoin), address(pip));
+
+        // auxiliary function added, not in tether source code
+        // basisPointsRate = 1, maximumFee = 1
+        usdt.changeFees(1, 1);
+
+        uint initbal = usdt.balanceOf(address(this));
```
Comment (kmbarry1) [inline src/DssDeploy.t.sol]: Variable is unused--remove or use?

File: src/tokens.sol [hunk 11]
```diff
@@ -666,3 +666,129 @@ contract MANA {
...
+    function balanceOf(address _owner) public view returns (uint balance) {
+        return balances[_owner];
+    }
+
+    function transfer(address _to, uint _value) public onlyPayloadSize(2 * 32) {
+        uint fee = (_value.mul(basisPointsRate)).div(10000);
+        if (fee > maximumFee) {
+            fee = maximumFee;
+        }
+        uint sendAmount = _value.sub(fee);
```
Comment (kmbarry1) [inline src/tokens.sol:744]: FYI this implementation of fees doesn't respect `decimals`, but maybe that doesn't matter for testing...

File: src/DssDeploy.t.sol (hunk 10 above)
Comment (wilbarnes) [reply src/DssDeploy.t.sol]: removed

File: src/tokens.sol (hunk 11 above)
Comment (wilbarnes) [reply src/tokens.sol:744]: well it is just basisPointsRate right which if it's a 1bp fee, then:

a `1000000` value txn would incur a fee of `100`

`1000000 * 1 / 10000 = 100`

and `100 / 1000000 = 0.0001` so one bp

Comment (kmbarry1) [reply src/tokens.sol:744]: Look at your test, if the maximum fee is set less than 1 basis point equivalent, it uses that instead.

Comment (kmbarry1) [reply src/tokens.sol:744]: `        assertEq(exitbal, 99999998);`
`99999998` is 99.999998 USDT, not 0.99999998

Comment (wilbarnes) [reply src/tokens.sol:744]: ah right fixed

**Review by kmbarry1 — APPROVED** [review]
LGTM

**Comment by gbalabasquer** [comment]
> two of the GemJoin6 tests, `testFailGemJoin6Join()` & `testFailGemJoin6Exit()` were missing `deployKeepAuth()` and were not testing what they were intended to, so i added that function call

good catch

File: src/join.sol [hunk 12]
```diff
@@ -427,3 +428,74 @@ contract AuthGemJoin is LibNote {
...
+    function mul(uint x, uint y) internal pure returns (uint z) {
+        require(y == 0 || (z = x * y) / y == x, "GemJoin7/overflow");
+    }
+
+    function join(address urn, uint wad) public note {
+        require(live == 1, "GemJoin7/not-live");
+        require(implementations[gem.upgradedAddress()] == 1, "GemJoin7/implementation-invalid");
+        uint bal = gem.balanceOf(address(this));
+        gem.transferFrom(msg.sender, address(this), wad);
+        uint wadt = mul(gem.balanceOf(address(this)) - bal, 10 ** (18 - dec));
```
Comment (gbalabasquer) [inline src/join.sol]: I'd be extra safe and would use `sub` instead of `-` for the balance difference.

Comment (kmbarry1) [reply src/join.sol]: +1

Comment (wilbarnes) [reply src/join.sol]: fixed in latest commit

**Review by gbalabasquer — APPROVED** [review]
I think we are fine now with it. I'd just wait few days more just in case something about liq 1.2 comes and then we can squash and merge.
In the meanwhile this shouldn't stop making the kovan deployment.

**Comment by wilbarnes** [comment]
- [ ] make sure there are new flips

- [ ] redeploy the gemjoin7 adapter
